### ISSUE TITLE
user and group support

### ISSUE THREAD

**Issue by dsschult** [WIPACrepo/iceprod#44]
(imported from trac: [#847](http://code.icecube.wisc.edu/projects/icecube/ticket/847))

Support different users and groups.

Group types:
- admin (can administer sites)
- production
- default
- (maybe even separate into IceCube?/PINGU/HEX specific) 

Users:
- users may belong to more than one group, but each dataset must have a single group.
- Authentication:
  - local account
  - OAuth / other web identity system
  - ldap? (sounds hard with the distributed nature of iceprod2) 

Tracking:
- Track resource usage by group and by user.

**Comment by dsschult** [comment]
Thoughts on group structure:

```
/                                      - default group
/IceCube                               - IceCube group
/IceCube/SimProd                       - simulation production
/IceCube/SimProd/Base                  - production of basic datasets
/IceCube/SimProd/Systematics           - production of systematics datasets
/IceCube/Filtering                     - data filtering
/PINGU                                 - PINGU work
```

The tree structure allows us to divide resources over each main group, then divide the group's resources at the subgroup level, and so on.  The easy way to apply this is when submitting a dataset, to set `priority = group * subgroup * sub-subgroup`.  Note that not specifying a subgroup when they exist is the same as `1 - sum(subgroups)`, so the remaining priority left.

We can also easily mark a site as running only a certain group (prefix '/IceCube') and it will run the subgroups automatically.

**Comment by dsschult** [comment]
Ugly algorithm that works to produce the right group priorities:

```python
base_groups = {
'/a': 0.1,
'/b': 0.8,
'/b/a': 0.45,
'/b/b': 0.45,
'/c': 0.1,
}

groups = {}
for g in base_groups:
    parts = g.split('/')
    print g, parts
    prio = 1.0
    for i in range(2,len(parts)+1):
        print i, '/'.join(parts[:i])
        prio *= base_groups['/'.join(parts[:i])]
    subtract = 0.0
    for k in base_groups:
        if k.startswith(g) and len(k.split('/')) == len(parts)+1:
            subtract += base_groups[k]
    groups[g] = prio * (1.0 - subtract)
```
produces:
```python
In [84]: groups
Out[84]: 
{'/a': 0.1,
 '/b': 0.07999999999999999,
 '/b/a': 0.36000000000000004,
 '/b/b': 0.36000000000000004,
 '/c': 0.1}
```

**Comment by dsschult** [comment]
Add a filter that can be applied at the local site, leaving `base_groups` as a global setting:
```python
filter = [
'/b',
'/c',
]

intermediate_groups = {p for p in base_groups if any(True for f in filter if f in p)}
```

**Comment by dsschult** [comment]
Working being done in a [branch](/WIPACrepo/iceprod/tree/groups):

3e0e7a92dd5eb68bd212c710dc979e171c2f64fe
Add group support for datasets.  Includes priority changes, filtering of groups at a remote instance.

Still todo:
* Web interface to edit groups
* Group membership (match users to groups)
* Group selection on dataset submission
* Group roles / permissions
* User authentication

**Comment by dsschult** [comment]
implemented since 2.4 refactor
